### DISCUSSION TITLE
Why not bind objects to server actions?

### DISCUSSION THREAD

**Discussion by huangeddie** [vercel/next.js#68254] [App Router] [answered]
[This section](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#passing-additional-arguments) teaches about binding additional arguments to server actions.

Is it reasonable to bind whole objects as a means of passing data? It seems like it'd be easier to validate than using something like zod? What would be the issues if any?

Something like

```
'use server'
 
export async function updateUser(aUserObject, formData) {
  // ...
}
```

```
'use client'
 
import { updateUser } from './actions'
 
export function UserProfile() {
 const userObject = {
    ...
 }

  return (
    <form action={updateUserWithId.bind(null, userObject)}>
      <input type="text" name="name" />
      <button type="submit">Update User Name</button>
    </form>
  )
}
```

**Comment by Mr-Meshky** [comment, marked as answer]
When it comes to binding whole objects to functions for passing data, there are a few key things to consider:

Validation and Type Safety: Binding whole objects can make passing multiple pieces of data simpler, but it might make validating individual fields harder. Libraries like Zod provide a structured way to validate data, ensuring each field meets specific criteria (like required fields, data types). When you bind an entire object, you need to ensure all fields within the object are valid before using them, which might require additional checks.

Immutability and State Management: Objects passed by reference can be mutable, meaning their contents can change unexpectedly unless managed carefully. This can lead to unintended side effects or bugs if the object is modified unintentionally between the time it's bound and the time it's used.

Performance Concerns: Binding entire objects can impact performance, especially if these objects are large or if function binding happens frequently. This is more of a concern in performance-sensitive applications or scenarios with high-frequency updates.

Code Maintainability: While binding objects simplifies passing data, it can also obscure the intent of the function call. Developers reading the code might find it less clear compared to passing individual parameters explicitly, especially if the function is only supposed to handle specific properties of the object.

In your example:

```
<form action={updateUserWithId.bind(null, userObject)}>
  <input type="text" name="name" />
  <button type="submit">Update User Name</button>
</form>
```

Here, updateUserWithId is bound with userObject, which might work fine depending on how updateUserWithId is implemented and how userObject is used within it. Just make sure that updateUserWithId handles userObject correctly and that any necessary validation and handling of its properties are done properly within updateUserWithId.

To sum up, while binding whole objects can be convenient for passing data, it’s important to balance convenience with considerations like validation, immutability, performance, and code clarity. Depending on the specific context and requirements of your app, using libraries like Zod for validation may still be necessary to ensure robustness and maintainability.

**Reply by huangeddie** [reply]
Thank you so much for that response. That was very helpful.

**Reply by huangeddie** [reply]
Also just wondering, did you happen to write that with an LLM? Your writing style reminded me a little of it

**Reply by Mr-Meshky** [reply]
Yes, because my English is very weak, I asked him to translate from Persian to English 😂
